### ISSUE TITLE
UDP/rsyslog: Invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by gdm85** [mozilla-services/heka#1439]
I have been trying an unusual configuration (heka 0.9.0):

```
[syslog_udp]
type = "UdpInput"
net = "unixgram"
address = "/dev/log"
decoder = "RsyslogDecoder"

[RstEncoder]

[LogOutput]
message_matcher = "TRUE"
encoder = "RstEncoder"
```

The runtime error is this:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x48a016]

goroutine 10 [running]:
github.com/mozilla-services/heka/pipeline.(*iRunner).Deliver(0xc2080a17a0, 0xc2080960f0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:438 +0x86
github.com/mozilla-services/heka/pipeline.(*sRunner).DeliverRecord(0xc2080404d0, 0xc208d6c000, 0x3e, 0x10000, 0x0, 0x0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/splitter_runner.go:256 +0x32c
github.com/mozilla-services/heka/pipeline.(*sRunner).SplitStream(0xc2080404d0, 0x7f7e5f825660, 0xc20802c1f0, 0x0, 0x0, 0x0, 0x0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/splitter_runner.go:315 +0x2e4
github.com/mozilla-services/heka/plugins/udp.(*UdpInput).Run(0xc20809e5a0, 0x7f7e5f81fd20, 0xc2080a17a0, 0x7f7e5f820080, 0xc208034680, 0x0, 0x0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/plugins/udp/udp_input.go:122 +0x43d
github.com/mozilla-services/heka/pipeline.(*iRunner).Starter(0xc2080a17a0, 0x7f7e5f820080, 0xc208034680, 0xc2080347c0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:283 +0x180
created by github.com/mozilla-services/heka/pipeline.(*iRunner).Start
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:266 +0x606

goroutine 1 [chan receive]:
github.com/mozilla-services/heka/pipeline.Run(0xc208034680)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/pipeline_runner.go:271 +0x143c
main.main()
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/cmd/hekad/main.go:191 +0x12a6

goroutine 5 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
	/usr/local/go/src/os/signal/signal_unix.go:27 +0x35

goroutine 6 [chan receive]:
github.com/mozilla-services/heka/plugins.(*LogOutput).Run(0xf9e9a0, 0x7f7e5f81ffc8, 0xc20800d340, 0x7f7e5f820080, 0xc208034680, 0x0, 0x0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/plugins/log_output.go:45 +0x1ca
github.com/mozilla-services/heka/pipeline.(*foRunner).Starter(0xc20800d340, 0x7f7e5f820080, 0xc208034680, 0xc2080ba5c0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:882 +0x7dd
created by github.com/mozilla-services/heka/pipeline.(*foRunner).Start
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:771 +0x66c

goroutine 7 [chan receive]:
github.com/mozilla-services/heka/pipeline.(*DiagnosticTracker).Run(0xc20809e690)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/packet_tracking.go:136 +0x195
created by github.com/mozilla-services/heka/pipeline.Run
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/pipeline_runner.go:252 +0xdf4

goroutine 8 [chan receive]:
github.com/mozilla-services/heka/pipeline.(*DiagnosticTracker).Run(0xc20809e6c0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/packet_tracking.go:136 +0x195
created by github.com/mozilla-services/heka/pipeline.Run
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/pipeline_runner.go:253 +0xe11

goroutine 9 [select]:
github.com/mozilla-services/heka/pipeline.func·017()
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/router.go:120 +0xa63
created by github.com/mozilla-services/heka/pipeline.(*messageRouter).Start
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/router.go:192 +0x93

goroutine 11 [chan receive]:
github.com/mozilla-services/heka/pipeline.func·019()
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/router.go:278 +0xf8
created by github.com/mozilla-services/heka/pipeline.(*MatchRunner).Start
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/router.go:315 +0x11c
```

Is this a bug or expected behavior?

**Comment by rafrombrc** [comment]
It's sort of both a bug *and* expected behavior. Your config not working is expected behavior, since you're trying to use a decoder called "RsyslogDecoder" but you haven't defined a decoder with that name. Getting a panic is a bug, though, b/c the correct behavior would be for Heka to just emit an error message to that effect and exit cleanly. You didn't include any of the output prior to the panic, but I'm guessing there was a line right before the panic that said `Input 'syslog_udp' error: decoder 'RsyslogDecoder' not registered`.

I'll work on fixing the panic. In the meantime, though, you should be able to work around this issue by adding a config section like the following:

```ini
[RsyslogDecoder]
type = "SandboxDecoder"
filename = "lua_decoders/rsyslog.lua"

[RsyslogDecoder.config]
type = "RSYSLOG_TraditionalFileFormat"
template = '%TIMESTAMP% %HOSTNAME% %syslogtag%%msg:::sp-if-no-1st-sp%%msg:::drop-last-lf%\n'
tz = "America/Los_Angeles"
```

The details of what should be in the `RsyslogDecoder` section varies based on your exact requirements, of course, but you have to have *something* there or you'll see the panic.

**Comment by gdm85** [comment]
I understand. Somehow I thought that was a pre-defined decoder.

**Comment by rafrombrc** [comment]
It's provided by Heka, but for it to be useful it requires an rsyslog template to be provided so it knows the exact format the input is in, so it knows how to parse it correctly. The only decoder that's automatically made available by Heka is `[ProtobufDecoder]`, any others need to be specified in the config.

Anyway, closing this issue since the panic has been fixed as of #1451.
